dev/io.openliberty.netty.internal.impl/src/io/openliberty/netty/internal/impl/NettyFrameworkImpl.java

```diff
@@ -0,0 +1,359 @@
+     * @throws Exception
+     */
+    public <T> T runWhenServerStarted(Callable<T> callable) throws Exception {
+        synchronized (syncStarted) {
+            if (!serverCompletelyStarted.get()) {
+                serverStartedTasks.add(callable);
+                return null;
+            }
+        }
+        return callable.call();
+    }
+
+    /*
+     * If the server has not completely started, then wait until it has been.
+     * The server will be "completely" stated when the server start signal has been received and any task waiting on that signal before running have now been run.
+     */
+    @FFDCIgnore({ InterruptedException.class })
+    public static void waitServerCompletelyStarted() {
+        synchronized (syncStarted) {
+            if (serverCompletelyStarted.get() == false) {
+                try {
+                    syncStarted.wait();
+                } catch (InterruptedException x) {
+                    // assume we can go one then
+                }
+            }
+        }
+        return;
+    }
+
+    /**
+     * non-blocking method to return the state of server startup with respect to the server being completely started.
+     * The server will be "completely" stated when the server start signal has been received and any task waiting on that signal before running have now been run.
+     *
+     * @return
+     */
+    public static boolean isServerCompletelyStarted() {
+        return serverCompletelyStarted.get();
+    }
+
+    /**
+     * Declarative Services method for unsetting the ServerStarted service
+     *
+     * @param ref reference to the service
+     */
+    protected synchronized void unsetServerStarted(ServiceReference<ServerStarted> ref) {
+        // server is shutting down
+    }
+    
+    @Override
+    public ServerBootstrapExtended createTCPBootstrap(Map<String, Object> tcpOptions) throws NettyException {
+        return TCPUtils.createTCPBootstrap(this, tcpOptions);
+    }
+
+    @Override
+    public BootstrapExtended createUDPBootstrap(Map<String, Object> options) throws NettyException {
+        return UDPUtils.createUDPBootstrap(this, options);
+    }
+
+
+    @Override
+    public ChannelFuture start(ServerBootstrapExtended bootstrap, String inetHost, int inetPort, ChannelFutureListener bindListener) throws NettyException {
+        return TCPUtils.start(this, bootstrap, inetHost, inetPort, bindListener);
+    }
+
+    @Override
+    public ChannelFuture start(BootstrapExtended bootstrap, String inetHost, int inetPort, ChannelFutureListener bindListener) throws NettyException {
+        return UDPUtils.start(this, bootstrap, inetHost, inetPort, bindListener);
+    }
+
+    @Override
+    public ChannelFuture stop(Channel channel) {
+        if (isActive && channel.isOpen()) {
+            ChannelFuture closeFuture = channel.close();
+            // TODO: should we wait until close actually completes to remove?
```

Comment:
If we did the server stop would send a second close as it loops round the activeChannels in serverStopping without an obvious equivalent of 
```if (isActive && channel.isOpen())``` 
in the serverStopping loop for any channel that was 'slow' to close and thus closing when the server was stopped.

Reply:
As we discussed - this logic isn't currently _wrong_, but we'll probably need to improve this area in the future. I've opened https://github.com/OpenLiberty/open-liberty/issues/18518